File: code97.py
class Solution(object):
    def isInterleave(self, s1, s2, s3):
        self.map=dict()
        if len(s1)+len(s2) != len(s3):
            return False
        else:
            return self.checklInterleave(s1, s2, s3, 0, 0, 0)

    def checklInterleave(self, s1, s2, s3, i1, i2, i3):
        print(s1[i1:],s2[i2:],s3[i3:])
        if len(s1) == i1 and len(s2) == i2 and len(s3) == i3:
            return True
        if i1<len(s1) and s1[i1]==s3[i3] and self.checklInterleave(s1,s2,s3,i1+1,i2,i3+1):
            return True
        elif i2<len(s2) and s2[i2]==s3[i3] and self.checklInterleave(s1,s2,s3,i1,i2+1,i3+1):
            return True
        else:
            return False
        
s=Solution()
s1='a'
s2=''
s3='a'
print(s.isInterleave(s1,s2,s3))
        

File: code97.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    map<pair<pair<int,int>,int>,int> dp;
    bool isInterleave(string s1, string s2, string s3) {
        if(s1.size()+s2.size()!=s3.size()) return false;
        return checkInterleave(s1,s2,s3,0,0,0);
    }

    bool checkInterleave(const string& s1,const string& s2,const string& s3, int i1,int i2,int i3) {
        if(s1.size()==i1 && s2.size()==i2 && s3.size()==i3){
            return true;
        }
        pair<int,int> conpair;
        conpair.first=i1;
        conpair.second=i2;
        pair<pair<int,int>,int> constractor_result_pair;
        constractor_result_pair.first=conpair;
        constractor_result_pair.second=i3;
        if(dp[constractor_result_pair]!=0){
            return dp[constractor_result_pair]==1;
        }
        if(i1<s1.size() && s1[i1]==s3[i3] && checkInterleave(s1,s2,s3,i1+1,i2,i3+1)){
            dp[constractor_result_pair]=1;
            return true;
        }
        if(i2<s2.size() && s2[i2]==s3[i3] && checkInterleave(s1,s2,s3,i1,i2+1,i3+1)){
            dp[constractor_result_pair]=1;
            return true;
        }
        else{
            dp[constractor_result_pair]=-1;
            return false;
        }

    }
};

int main() {
    string s1="aabcc";
    string s2="dbbca", s3 = "aadbbbaccc";
    Solution s;
    cout<<s.isInterleave(s1,s2,s3)<<endl;
    return 0;

}